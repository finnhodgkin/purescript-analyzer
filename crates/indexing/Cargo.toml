[package]
name = "indexing"
version = "0.1.0"
edition = "2021"

[dependencies]
fxhash = "0.2.1"
indexmap = "2.7.0"
la-arena = "0.3.1"
petgraph = "0.7.0"
rowan = "0.16.1"
rustc-hash = "2.1.0"
smol_str = "0.3.2"
syntax = { version = "0.1.0", path = "../syntax" }

[dev-dependencies]
lexing = { path = "../lexing" }
parsing = { path = "../parsing" }
