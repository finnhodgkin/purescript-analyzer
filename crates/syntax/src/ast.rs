#[macro_use]
mod macros;

mod binder;
mod binding;
mod common;
mod declaration;
mod module;
mod name;

pub use binder::*;
pub use binding::*;
pub use common::*;
pub use declaration::*;
pub use module::*;
pub use name::*;
