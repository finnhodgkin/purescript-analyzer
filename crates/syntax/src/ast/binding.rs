use rowan::ast::AstNode;

use super::OneOrMore;

_create_ast_v!(Binding, UnconditionalBinding(UnconditionalBinding), GuardedBinding(GuardedBinding));

_create_ast!(GuardedExpression, WhereExpression);

impl UnconditionalBinding {
    pub fn where_expression(&self) -> Option<WhereExpression> {
        WhereExpression::cast(self.node.first_child()?)
    }
}

impl GuardedBinding {
    pub fn guarded_expressions(&self) -> Option<OneOrMore<GuardedExpression>> {
        OneOrMore::cast(self.node.first_child()?)
    }
}

impl GuardedExpression {
    pub fn where_expression(&self) -> Option<WhereExpression> {
        WhereExpression::cast(self.node.last_child()?)
    }
}
