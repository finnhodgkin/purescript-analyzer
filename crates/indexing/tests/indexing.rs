use indexing::{
    Duplicate, ExprItem, FullIndexingResult, Id, IndexingError, TypeGroupId, TypeItem, ValueGroupId,
};
use rowan::ast::AstNode;
use syntax::cst;

fn index<'line>(source: impl AsRef<[&'line str]>) -> (cst::Module, FullIndexingResult) {
    let source = format!("module Main where\n{}", source.as_ref().join("\n"));

    let lexed = lexing::lex(&source);
    let tokens = lexing::layout(&lexed);
    let (node, _) = parsing::parse(&lexed, &tokens);

    let module = cst::Module::cast(node).unwrap();
    let index = indexing::index(&module);

    (module, index)
}

#[test]
fn valid_module() {
    let (module, index) = index(&[
        "life :: Int",
        "life = 42",
        "data Boolean :: Type",
        "data Boolean = True | False",
        "newtype Id :: Type -> Type",
        "newtype Id a = Id a",
        "type Fn :: Type -> Type -> Type",
        "type Fn a b = a -> b",
        "class Eq :: Type -> Constraint",
        "class Eq a where",
        "  eq :: a -> a -> Boolean",
        "instance eqInt :: Eq Int where",
        "  eq = eqIntImpl",
    ]);

    assert!(index.errors.is_empty());

    assert!(matches!(index.nominal.lookup_type_item("Boolean"), Some((_, TypeItem::Data(_)))));
    assert!(matches!(index.nominal.lookup_type_item("Id"), Some((_, TypeItem::Newtype(_)))));
    assert!(matches!(index.nominal.lookup_type_item("Fn"), Some((_, TypeItem::Synonym(_)))));
    assert!(matches!(index.nominal.lookup_type_item("Eq"), Some((_, TypeItem::Class(_)))));

    if let Some((_, ExprItem::Value(ValueGroupId { signature: Some(id), .. }))) =
        index.nominal.lookup_expr_item("life")
    {
        let ptr = index.source_map.declaration_ptr(*id).unwrap();
        let node = ptr.to_node(module.syntax());

        assert_eq!(node.syntax().text(), "\nlife :: Int");
        assert_eq!(index.source_map.declaration_id(ptr), Some(*id));
    } else {
        unreachable!()
    }

    if let Some((_, ExprItem::Method(id))) = index.nominal.lookup_expr_item("eq") {
        let ptr = index.source_map.class_member_ptr(*id).unwrap();
        let node = ptr.to_node(module.syntax());

        assert_eq!(node.syntax().text(), "\n  eq :: a -> a -> Boolean");
        assert_eq!(index.source_map.class_member_id(ptr), Some(*id));
    } else {
        unreachable!()
    }

    if let Some((_, ExprItem::Instance(id))) = index.nominal.lookup_expr_item("eqInt") {
        let ptr = index.source_map.instance_ptr(*id).unwrap();
        let node = ptr.to_node(module.syntax());

        assert_eq!(node.syntax().text(), "\ninstance eqInt :: Eq Int where\n  eq = eqIntImpl");
        assert_eq!(index.source_map.instance_id(ptr), Some(*id));
    } else {
        unreachable!()
    }

    if let Some((_, ExprItem::Constructor(id))) = index.nominal.lookup_expr_item("True") {
        let ptr = index.source_map.constructor_ptr(*id).unwrap();
        let node = ptr.to_node(module.syntax());

        assert_eq!(node.syntax().text(), " True");
        assert_eq!(index.source_map.constructor_id(ptr), Some(*id));
    } else {
        unreachable!()
    }

    if let Some((_, ExprItem::Constructor(id))) = index.nominal.lookup_expr_item("Id") {
        if let Some(id) = index.relational.of_constructor(*id) {
            assert!(matches!(index.nominal.index_type_item(id), Some(TypeItem::Newtype(_))));
        } else {
            unreachable!()
        }
    } else {
        unreachable!()
    }

    if let Some((_, ExprItem::Method(id))) = index.nominal.lookup_expr_item("eq") {
        if let Some(id) = index.relational.of_method(*id) {
            assert!(matches!(index.nominal.index_type_item(id), Some(TypeItem::Class(_))));
        } else {
            unreachable!()
        }
    } else {
        unreachable!()
    }

    let (id, item) = index.nominal.lookup_expr_item("life").unwrap();
    assert_eq!(index.nominal.index_expr_item(id), Some(item));

    if let Some((id, _)) = index.nominal.lookup_type_item("Boolean") {
        let constructors: Vec<_> = index.relational.constructors_of(id).collect();
        if let &[t, f] = &constructors[..] {
            let t_ptr = index.source_map.constructor_ptr(t).unwrap();
            let t_node = t_ptr.to_node(module.syntax());

            assert_eq!(t_node.syntax().text(), " True");
            assert_eq!(index.source_map.constructor_id(t_ptr), Some(t));

            let f_ptr = index.source_map.constructor_ptr(f).unwrap();
            let f_node = f_ptr.to_node(module.syntax());

            assert_eq!(f_node.syntax().text(), " False");
            assert_eq!(index.source_map.constructor_id(f_ptr), Some(f));
        } else {
            unreachable!()
        }
    } else {
        unreachable!()
    }

    if let Some((id, _)) = index.nominal.lookup_type_item("Id") {
        let constructors: Vec<_> = index.relational.constructors_of(id).collect();
        if let &[id] = &constructors[..] {
            let ptr = index.source_map.constructor_ptr(id).unwrap();
            let node = ptr.to_node(module.syntax());

            assert_eq!(node.syntax().text(), " Id a");
            assert_eq!(index.source_map.constructor_id(ptr), Some(id));
        } else {
            unreachable!()
        }
    } else {
        unreachable!()
    }

    if let Some((id, _)) = index.nominal.lookup_type_item("Eq") {
        let methods: Vec<_> = index.relational.methods_of(id).collect();
        if let &[id] = &methods[..] {
            let ptr = index.source_map.class_member_ptr(id).unwrap();
            let node = ptr.to_node(module.syntax());

            assert_eq!(node.syntax().text(), "\n  eq :: a -> a -> Boolean");
            assert_eq!(index.source_map.class_member_id(ptr), Some(id));
        } else {
            unreachable!()
        }
    } else {
        unreachable!()
    }

    if let Some((_, TypeItem::Synonym(TypeGroupId { declaration: Some(id), .. }))) =
        index.nominal.lookup_type_item("Fn")
    {
        let ptr = index.source_map.declaration_ptr(*id).unwrap();
        let node = ptr.to_node(module.syntax());

        assert_eq!(node.syntax().text(), "\ntype Fn a b = a -> b");
        assert_eq!(index.source_map.declaration_id(ptr), Some(*id));
    } else {
        unreachable!()
    }
}

#[test]
fn late_signature_error() {
    let (_, index) = index(&[
        "life = 42",
        "life :: Int",
        "data Boolean = True | False",
        "data Boolean :: Type",
        "newtype Id a = Id a",
        "newtype Id :: Type -> Type",
        "type Fn a b = a -> b",
        "type Fn :: Type -> Type -> Type",
        "class Eq a",
        "class Eq :: Type -> Constraint",
    ]);
    assert_eq!(
        &index.errors,
        &[
            IndexingError::LateSignature {
                declaration: Id::from_raw(0),
                signature: Id::from_raw(1)
            },
            IndexingError::LateSignature {
                declaration: Id::from_raw(2),
                signature: Id::from_raw(3)
            },
            IndexingError::LateSignature {
                declaration: Id::from_raw(4),
                signature: Id::from_raw(5)
            },
            IndexingError::LateSignature {
                declaration: Id::from_raw(6),
                signature: Id::from_raw(7)
            },
            IndexingError::LateSignature {
                declaration: Id::from_raw(8),
                signature: Id::from_raw(9)
            },
        ]
    );
}

#[test]
fn duplicate_expr_item() {
    let (_, index) = index(&[
        "life :: Int",
        "life = 42",
        "instance life :: Life",
        "class Life where",
        "  life :: Int",
        "life :: Int",
        "data IdData a = Id a",
        "newtype IdNewtype a = Id a",
    ]);
    assert_eq!(
        &index.errors,
        &[
            IndexingError::DuplicateExprItem {
                item_id: Id::from_raw(0),
                duplicate: Duplicate::Instance(Id::from_raw(0)),
            },
            IndexingError::DuplicateExprItem {
                item_id: Id::from_raw(0),
                duplicate: Duplicate::ClassMember(Id::from_raw(0)),
            },
            IndexingError::LateSignature {
                declaration: Id::from_raw(1),
                signature: Id::from_raw(4)
            },
            IndexingError::DuplicateExprItem {
                item_id: Id::from_raw(0),
                duplicate: Duplicate::Declaration(Id::from_raw(4)),
            },
            IndexingError::DuplicateExprItem {
                item_id: Id::from_raw(1),
                duplicate: Duplicate::Constructor(Id::from_raw(1)),
            },
        ]
    );
}

#[test]
fn value_is_duplicate() {
    let (_, index) = index([
        "class Eq a where",
        "  eq :: a -> a -> Boolean",
        "eq :: Int -> Int -> Boolean",
        "eq = eqIntImpl",
    ]);
    assert_eq!(
        &index.errors,
        &[
            IndexingError::DuplicateExprItem {
                item_id: Id::from_raw(0),
                duplicate: Duplicate::Declaration(Id::from_raw(1)),
            },
            IndexingError::DuplicateExprItem {
                item_id: Id::from_raw(0),
                duplicate: Duplicate::Declaration(Id::from_raw(2)),
            }
        ]
    );
}

#[test]
fn duplicate_data() {
    let (_, index) =
        index(&["data Id0 :: Type", "data Id0 :: Type", "data Id1 = Id2", "data Id1 = Id3"]);
    assert_eq!(
        &index.errors,
        &[
            IndexingError::DuplicateTypeItem {
                item_id: Id::from_raw(0),
                duplicate: Duplicate::Declaration(Id::from_raw(1)),
            },
            IndexingError::DuplicateTypeItem {
                item_id: Id::from_raw(1),
                duplicate: Duplicate::Declaration(Id::from_raw(3)),
            },
        ]
    );
}

#[test]
fn duplicate_newtype() {
    let (_, index) = index(&[
        "newtype Id0 :: Type",
        "newtype Id0 :: Type",
        "newtype Id1 = Id2 Int",
        "newtype Id1 = Id3 Int",
    ]);
    assert_eq!(
        &index.errors,
        &[
            IndexingError::DuplicateTypeItem {
                item_id: Id::from_raw(0),
                duplicate: Duplicate::Declaration(Id::from_raw(1)),
            },
            IndexingError::DuplicateTypeItem {
                item_id: Id::from_raw(1),
                duplicate: Duplicate::Declaration(Id::from_raw(3)),
            },
        ]
    );
}

#[test]
fn duplicate_class() {
    let (_, index) = index(&["class Id0 :: Type", "class Id0 :: Type", "class Id1", "class Id1"]);
    assert_eq!(
        &index.errors,
        &[
            IndexingError::DuplicateTypeItem {
                item_id: Id::from_raw(0),
                duplicate: Duplicate::Declaration(Id::from_raw(1)),
            },
            IndexingError::DuplicateTypeItem {
                item_id: Id::from_raw(1),
                duplicate: Duplicate::Declaration(Id::from_raw(3)),
            },
        ]
    );
}

#[test]
fn duplicate_synonym() {
    let (_, index) =
        index(&["type Id0 :: Type", "type Id0 :: Type", "type Id1 = Id0", "type Id1 = Id0"]);
    assert_eq!(
        &index.errors,
        &[
            IndexingError::DuplicateTypeItem {
                item_id: Id::from_raw(0),
                duplicate: Duplicate::Declaration(Id::from_raw(1)),
            },
            IndexingError::DuplicateTypeItem {
                item_id: Id::from_raw(1),
                duplicate: Duplicate::Declaration(Id::from_raw(3)),
            },
        ]
    );
}

#[test]
fn duplicate_type_item() {
    let (_, index) = index(&[
        "data Id0 :: Type",
        "data Id0 = Id10 Int",
        "newtype Id0 :: Type",
        "newtype Id0 = Id11 Int",
        "class Id0 :: Type",
        "class Id0",
        "type Id0 :: Type",
        "type Id0 = Id0",
    ]);
    assert_eq!(
        &index.errors,
        &[
            IndexingError::DuplicateTypeItem {
                item_id: Id::from_raw(0),
                duplicate: Duplicate::Declaration(Id::from_raw(2)),
            },
            IndexingError::DuplicateTypeItem {
                item_id: Id::from_raw(0),
                duplicate: Duplicate::Declaration(Id::from_raw(3)),
            },
            IndexingError::DuplicateTypeItem {
                item_id: Id::from_raw(0),
                duplicate: Duplicate::Declaration(Id::from_raw(4)),
            },
            IndexingError::DuplicateTypeItem {
                item_id: Id::from_raw(0),
                duplicate: Duplicate::Declaration(Id::from_raw(5)),
            },
            IndexingError::DuplicateTypeItem {
                item_id: Id::from_raw(0),
                duplicate: Duplicate::Declaration(Id::from_raw(6)),
            },
            IndexingError::DuplicateTypeItem {
                item_id: Id::from_raw(0),
                duplicate: Duplicate::Declaration(Id::from_raw(7)),
            },
        ]
    );
}

#[test]
fn data_is_duplicate() {
    let (_, index) = index(&["class Eq", "data Eq :: Type", "data Eq"]);
    assert_eq!(
        &index.errors,
        &[
            IndexingError::DuplicateTypeItem {
                item_id: Id::from_raw(0),
                duplicate: Duplicate::Declaration(Id::from_raw(1)),
            },
            IndexingError::DuplicateTypeItem {
                item_id: Id::from_raw(0),
                duplicate: Duplicate::Declaration(Id::from_raw(2)),
            }
        ]
    );
}
