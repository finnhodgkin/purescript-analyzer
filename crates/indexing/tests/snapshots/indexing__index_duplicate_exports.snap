---
source: crates/indexing/tests/indexing.rs
expression: snapshot
snapshot_kind: text
---
Index {
    has_exports: true,
    term_item: Arena {
        len: 10,
        data: [
            Value {
                signature: Some(
                    Idx::<ValueSignature>>(0),
                ),
                equations: [
                    Idx::<ValueEquation>>(0),
                ],
            },
            Constructor {
                id: Idx::<DataConstructor>>(0),
            },
            Constructor {
                id: Idx::<DataConstructor>>(1),
            },
            Constructor {
                id: Idx::<DataConstructor>>(2),
            },
            ClassMember {
                id: Idx::<ClassMemberStatement>>(0),
            },
            Instance {
                id: Idx::<InstanceDeclaration>>(0),
            },
            Instance {
                id: Idx::<InstanceDeclaration>>(1),
            },
            Derive {
                id: Idx::<DeriveDeclaration>>(0),
            },
            Foreign {
                id: Idx::<ForeignImportValueDeclaration>>(0),
            },
            Operator {
                id: Idx::<InfixDeclaration>>(0),
            },
        ],
    },
    type_item: Arena {
        len: 6,
        data: [
            Data {
                signature: Some(
                    Idx::<DataSignature>>(0),
                ),
                equation: Some(
                    Idx::<DataEquation>>(0),
                ),
                role: None,
            },
            Newtype {
                signature: Some(
                    Idx::<NewtypeSignature>>(0),
                ),
                equation: Some(
                    Idx::<NewtypeEquation>>(0),
                ),
                role: Some(
                    Idx::<TypeRoleDeclaration>>(0),
                ),
            },
            Synonym {
                signature: Some(
                    Idx::<TypeSynonymSignature>>(0),
                ),
                equation: Some(
                    Idx::<TypeSynonymEquation>>(0),
                ),
            },
            Class {
                signature: Some(
                    Idx::<ClassSignature>>(0),
                ),
                declaration: Some(
                    Idx::<ClassDeclaration>>(0),
                ),
            },
            Foreign {
                id: Idx::<ForeignImportDataDeclaration>>(0),
            },
            Operator {
                id: Idx::<InfixDeclaration>>(1),
            },
        ],
    },
    term_export: {
        Idx::<TermItem>(0): Idx::<ExportItem>>(0),
        Idx::<TermItem>(9): Idx::<ExportItem>>(14),
        Idx::<TermItem>(3): Idx::<ExportItem>>(8),
        Idx::<TermItem>(1): Idx::<ExportItem>>(6),
        Idx::<TermItem>(8): Idx::<ExportItem>>(2),
        Idx::<TermItem>(2): Idx::<ExportItem>>(6),
    },
    type_export: {
        Idx::<TypeItem>(0): Idx::<ExportItem>>(6),
        Idx::<TypeItem>(3): Idx::<ExportItem>>(4),
        Idx::<TypeItem>(1): Idx::<ExportItem>>(8),
        Idx::<TypeItem>(5): Idx::<ExportItem>>(16),
        Idx::<TypeItem>(4): Idx::<ExportItem>>(12),
        Idx::<TypeItem>(2): Idx::<ExportItem>>(10),
    },
    import_name: {
        Idx::<ImportStatement>>(0): "Prelude",
        Idx::<ImportStatement>>(3): "Data.Map",
        Idx::<ImportStatement>>(1): "Data.List",
        Idx::<ImportStatement>>(2): "Data.Set",
    },
    import_items: {
        Idx::<ImportStatement>>(0): ImportedItems {
            terms: {},
            types: {},
            exported: false,
        },
        Idx::<ImportStatement>>(3): ImportedItems {
            terms: {},
            types: {},
            exported: true,
        },
        Idx::<ImportStatement>>(1): ImportedItems {
            terms: {},
            types: {},
            exported: false,
        },
        Idx::<ImportStatement>>(2): ImportedItems {
            terms: {},
            types: {},
            exported: true,
        },
    },
    alias_nominal: {
        "Exports": [
            Idx::<ImportStatement>>(2),
            Idx::<ImportStatement>>(3),
        ],
        "List": [
            Idx::<ImportStatement>>(1),
        ],
    },
    term_nominal: {
        "+": Idx::<TermItem>(9),
        "unit": Idx::<TermItem>(8),
        "eqId": Idx::<TermItem>(7),
        "eq": Idx::<TermItem>(4),
        "id": Idx::<TermItem>(0),
        "Just": Idx::<TermItem>(1),
        "Id": Idx::<TermItem>(3),
        "Nothing": Idx::<TermItem>(2),
        "eqInt": Idx::<TermItem>(5),
    },
    type_nominal: {
        "MaybeId": Idx::<TypeItem>(2),
        "Unit": Idx::<TypeItem>(4),
        "Id": Idx::<TypeItem>(1),
        "Eq": Idx::<TypeItem>(3),
        "Maybe": Idx::<TypeItem>(0),
        "+": Idx::<TypeItem>(5),
    },
}
Relational {
    data: [
        (
            Idx::<TypeItem>(0),
            Idx::<TermItem>(1),
        ),
        (
            Idx::<TypeItem>(0),
            Idx::<TermItem>(2),
        ),
        (
            Idx::<TypeItem>(1),
            Idx::<TermItem>(3),
        ),
    ],
    class: [
        (
            Idx::<TypeItem>(3),
            Idx::<TermItem>(4),
        ),
    ],
    chain: [
        (
            Idx::<InstanceChain>>(0),
            Idx::<InstanceDeclaration>>(0),
        ),
        (
            Idx::<InstanceChain>>(1),
            Idx::<InstanceDeclaration>>(1),
        ),
    ],
    instance: [
        (
            Idx::<InstanceDeclaration>>(0),
            Idx::<InstanceMemberStatement>>(0),
        ),
        (
            Idx::<InstanceDeclaration>>(1),
            Idx::<InstanceMemberStatement>>(1),
        ),
    ],
}
IndexingSource {
    export: AstPtrMap {
        arena: Arena {
            len: 20,
            data: [
                AstPtr {
                    raw: SyntaxNodePtr {
                        kind: ExportValue,
                        range: 16..19,
                    },
                },
                AstPtr {
                    raw: SyntaxNodePtr {
                        kind: ExportValue,
                        range: 23..26,
                    },
                },
                AstPtr {
                    raw: SyntaxNodePtr {
                        kind: ExportValue,
                        range: 30..35,
                    },
                },
                AstPtr {
                    raw: SyntaxNodePtr {
                        kind: ExportValue,
                        range: 39..44,
                    },
                },
                AstPtr {
                    raw: SyntaxNodePtr {
                        kind: ExportClass,
                        range: 48..57,
                    },
                },
                AstPtr {
                    raw: SyntaxNodePtr {
                        kind: ExportClass,
                        range: 61..70,
                    },
                },
                AstPtr {
                    raw: SyntaxNodePtr {
                        kind: ExportType,
                        range: 74..84,
                    },
                },
                AstPtr {
                    raw: SyntaxNodePtr {
                        kind: ExportType,
                        range: 88..98,
                    },
                },
                AstPtr {
                    raw: SyntaxNodePtr {
                        kind: ExportType,
                        range: 102..109,
                    },
                },
                AstPtr {
                    raw: SyntaxNodePtr {
                        kind: ExportType,
                        range: 113..120,
                    },
                },
                AstPtr {
                    raw: SyntaxNodePtr {
                        kind: ExportType,
                        range: 124..132,
                    },
                },
                AstPtr {
                    raw: SyntaxNodePtr {
                        kind: ExportType,
                        range: 136..144,
                    },
                },
                AstPtr {
                    raw: SyntaxNodePtr {
                        kind: ExportType,
                        range: 148..153,
                    },
                },
                AstPtr {
                    raw: SyntaxNodePtr {
                        kind: ExportType,
                        range: 157..162,
                    },
                },
                AstPtr {
                    raw: SyntaxNodePtr {
                        kind: ExportOperator,
                        range: 166..170,
                    },
                },
                AstPtr {
                    raw: SyntaxNodePtr {
                        kind: ExportOperator,
                        range: 174..178,
                    },
                },
                AstPtr {
                    raw: SyntaxNodePtr {
                        kind: ExportTypeOperator,
                        range: 182..191,
                    },
                },
                AstPtr {
                    raw: SyntaxNodePtr {
                        kind: ExportTypeOperator,
                        range: 195..204,
                    },
                },
                AstPtr {
                    raw: SyntaxNodePtr {
                        kind: ExportModule,
                        range: 208..223,
                    },
                },
                AstPtr {
                    raw: SyntaxNodePtr {
                        kind: ExportModule,
                        range: 227..242,
                    },
                },
            ],
        },
        table: {
            Idx::<ExportItem>>(19),
            Idx::<ExportItem>>(8),
            Idx::<ExportItem>>(10),
            Idx::<ExportItem>>(0),
            Idx::<ExportItem>>(13),
            Idx::<ExportItem>>(11),
            Idx::<ExportItem>>(12),
            Idx::<ExportItem>>(9),
            Idx::<ExportItem>>(3),
            Idx::<ExportItem>>(7),
            Idx::<ExportItem>>(17),
            Idx::<ExportItem>>(1),
            Idx::<ExportItem>>(18),
            Idx::<ExportItem>>(5),
            Idx::<ExportItem>>(2),
            Idx::<ExportItem>>(14),
            Idx::<ExportItem>>(4),
            Idx::<ExportItem>>(15),
            Idx::<ExportItem>>(16),
            Idx::<ExportItem>>(6),
        },
    },
    import: AstPtrMap {
        arena: Arena {
            len: 0,
            data: [],
        },
        table: {},
    },
    import_statement: AstPtrMap {
        arena: Arena {
            len: 4,
            data: [
                AstPtr {
                    raw: SyntaxNodePtr {
                        kind: ImportStatement,
                        range: 252..268,
                    },
                },
                AstPtr {
                    raw: SyntaxNodePtr {
                        kind: ImportStatement,
                        range: 268..293,
                    },
                },
                AstPtr {
                    raw: SyntaxNodePtr {
                        kind: ImportStatement,
                        range: 293..320,
                    },
                },
                AstPtr {
                    raw: SyntaxNodePtr {
                        kind: ImportStatement,
                        range: 320..347,
                    },
                },
            ],
        },
        table: {
            Idx::<ImportStatement>>(1),
            Idx::<ImportStatement>>(0),
            Idx::<ImportStatement>>(2),
            Idx::<ImportStatement>>(3),
        },
    },
    data_signature: AstPtrMap {
        arena: Arena {
            len: 1,
            data: [
                AstPtr {
                    raw: SyntaxNodePtr {
                        kind: DataSignature,
                        range: 380..408,
                    },
                },
            ],
        },
        table: {
            Idx::<DataSignature>>(0),
        },
    },
    data_equation: AstPtrMap {
        arena: Arena {
            len: 1,
            data: [
                AstPtr {
                    raw: SyntaxNodePtr {
                        kind: DataEquation,
                        range: 408..440,
                    },
                },
            ],
        },
        table: {
            Idx::<DataEquation>>(0),
        },
    },
    data_constructor: AstPtrMap {
        arena: Arena {
            len: 3,
            data: [
                AstPtr {
                    raw: SyntaxNodePtr {
                        kind: DataConstructor,
                        range: 423..430,
                    },
                },
                AstPtr {
                    raw: SyntaxNodePtr {
                        kind: DataConstructor,
                        range: 432..440,
                    },
                },
                AstPtr {
                    raw: SyntaxNodePtr {
                        kind: DataConstructor,
                        range: 483..490,
                    },
                },
            ],
        },
        table: {
            Idx::<DataConstructor>>(2),
            Idx::<DataConstructor>>(1),
            Idx::<DataConstructor>>(0),
        },
    },
    newtype_signature: AstPtrMap {
        arena: Arena {
            len: 1,
            data: [
                AstPtr {
                    raw: SyntaxNodePtr {
                        kind: NewtypeSignature,
                        range: 440..468,
                    },
                },
            ],
        },
        table: {
            Idx::<NewtypeSignature>>(0),
        },
    },
    newtype_equation: AstPtrMap {
        arena: Arena {
            len: 1,
            data: [
                AstPtr {
                    raw: SyntaxNodePtr {
                        kind: NewtypeEquation,
                        range: 468..490,
                    },
                },
            ],
        },
        table: {
            Idx::<NewtypeEquation>>(0),
        },
    },
    type_role: AstPtrMap {
        arena: Arena {
            len: 1,
            data: [
                AstPtr {
                    raw: SyntaxNodePtr {
                        kind: TypeRoleDeclaration,
                        range: 490..512,
                    },
                },
            ],
        },
        table: {
            Idx::<TypeRoleDeclaration>>(0),
        },
    },
    type_signature: AstPtrMap {
        arena: Arena {
            len: 1,
            data: [
                AstPtr {
                    raw: SyntaxNodePtr {
                        kind: TypeSynonymSignature,
                        range: 512..534,
                    },
                },
            ],
        },
        table: {
            Idx::<TypeSynonymSignature>>(0),
        },
    },
    type_equation: AstPtrMap {
        arena: Arena {
            len: 1,
            data: [
                AstPtr {
                    raw: SyntaxNodePtr {
                        kind: TypeSynonymEquation,
                        range: 534..558,
                    },
                },
            ],
        },
        table: {
            Idx::<TypeSynonymEquation>>(0),
        },
    },
    class_signature: AstPtrMap {
        arena: Arena {
            len: 1,
            data: [
                AstPtr {
                    raw: SyntaxNodePtr {
                        kind: ClassSignature,
                        range: 558..590,
                    },
                },
            ],
        },
        table: {
            Idx::<ClassSignature>>(0),
        },
    },
    class_declaration: AstPtrMap {
        arena: Arena {
            len: 1,
            data: [
                AstPtr {
                    raw: SyntaxNodePtr {
                        kind: ClassDeclaration,
                        range: 590..633,
                    },
                },
            ],
        },
        table: {
            Idx::<ClassDeclaration>>(0),
        },
    },
    class_member: AstPtrMap {
        arena: Arena {
            len: 1,
            data: [
                AstPtr {
                    raw: SyntaxNodePtr {
                        kind: ClassMemberStatement,
                        range: 607..633,
                    },
                },
            ],
        },
        table: {
            Idx::<ClassMemberStatement>>(0),
        },
    },
    value_signature: AstPtrMap {
        arena: Arena {
            len: 1,
            data: [
                AstPtr {
                    raw: SyntaxNodePtr {
                        kind: ValueSignature,
                        range: 347..371,
                    },
                },
            ],
        },
        table: {
            Idx::<ValueSignature>>(0),
        },
    },
    value_equation: AstPtrMap {
        arena: Arena {
            len: 1,
            data: [
                AstPtr {
                    raw: SyntaxNodePtr {
                        kind: ValueEquation,
                        range: 371..380,
                    },
                },
            ],
        },
        table: {
            Idx::<ValueEquation>>(0),
        },
    },
    chain: AstPtrMap {
        arena: Arena {
            len: 2,
            data: [
                AstPtr {
                    raw: SyntaxNodePtr {
                        kind: InstanceChain,
                        range: 633..682,
                    },
                },
                AstPtr {
                    raw: SyntaxNodePtr {
                        kind: InstanceChain,
                        range: 682..733,
                    },
                },
            ],
        },
        table: {
            Idx::<InstanceChain>>(1),
            Idx::<InstanceChain>>(0),
        },
    },
    instance: AstPtrMap {
        arena: Arena {
            len: 2,
            data: [
                AstPtr {
                    raw: SyntaxNodePtr {
                        kind: InstanceDeclaration,
                        range: 633..682,
                    },
                },
                AstPtr {
                    raw: SyntaxNodePtr {
                        kind: InstanceDeclaration,
                        range: 682..733,
                    },
                },
            ],
        },
        table: {
            Idx::<InstanceDeclaration>>(0),
            Idx::<InstanceDeclaration>>(1),
        },
    },
    instance_member: AstPtrMap {
        arena: Arena {
            len: 2,
            data: [
                AstPtr {
                    raw: SyntaxNodePtr {
                        kind: InstanceEquationStatement,
                        range: 665..682,
                    },
                },
                AstPtr {
                    raw: SyntaxNodePtr {
                        kind: InstanceEquationStatement,
                        range: 706..733,
                    },
                },
            ],
        },
        table: {
            Idx::<InstanceMemberStatement>>(1),
            Idx::<InstanceMemberStatement>>(0),
        },
    },
    derive: AstPtrMap {
        arena: Arena {
            len: 1,
            data: [
                AstPtr {
                    raw: SyntaxNodePtr {
                        kind: DeriveDeclaration,
                        range: 733..764,
                    },
                },
            ],
        },
        table: {
            Idx::<DeriveDeclaration>>(0),
        },
    },
    infix: AstPtrMap {
        arena: Arena {
            len: 2,
            data: [
                AstPtr {
                    raw: SyntaxNodePtr {
                        kind: InfixDeclaration,
                        range: 826..844,
                    },
                },
                AstPtr {
                    raw: SyntaxNodePtr {
                        kind: InfixDeclaration,
                        range: 844..868,
                    },
                },
            ],
        },
        table: {
            Idx::<InfixDeclaration>>(0),
            Idx::<InfixDeclaration>>(1),
        },
    },
    foreign_data: AstPtrMap {
        arena: Arena {
            len: 1,
            data: [
                AstPtr {
                    raw: SyntaxNodePtr {
                        kind: ForeignImportDataDeclaration,
                        range: 793..826,
                    },
                },
            ],
        },
        table: {
            Idx::<ForeignImportDataDeclaration>>(0),
        },
    },
    foreign_value: AstPtrMap {
        arena: Arena {
            len: 1,
            data: [
                AstPtr {
                    raw: SyntaxNodePtr {
                        kind: ForeignImportValueDeclaration,
                        range: 764..793,
                    },
                },
            ],
        },
        table: {
            Idx::<ForeignImportValueDeclaration>>(0),
        },
    },
}
[
    DuplicateExport {
        id: Idx::<ExportItem>>(1),
        existing: Idx::<ExportItem>>(0),
    },
    DuplicateExport {
        id: Idx::<ExportItem>>(3),
        existing: Idx::<ExportItem>>(2),
    },
    DuplicateExport {
        id: Idx::<ExportItem>>(5),
        existing: Idx::<ExportItem>>(4),
    },
    DuplicateExport {
        id: Idx::<ExportItem>>(7),
        existing: Idx::<ExportItem>>(6),
    },
    DuplicateExport {
        id: Idx::<ExportItem>>(7),
        existing: Idx::<ExportItem>>(6),
    },
    DuplicateExport {
        id: Idx::<ExportItem>>(7),
        existing: Idx::<ExportItem>>(6),
    },
    DuplicateExport {
        id: Idx::<ExportItem>>(9),
        existing: Idx::<ExportItem>>(8),
    },
    DuplicateExport {
        id: Idx::<ExportItem>>(9),
        existing: Idx::<ExportItem>>(8),
    },
    DuplicateExport {
        id: Idx::<ExportItem>>(11),
        existing: Idx::<ExportItem>>(10),
    },
    DuplicateExport {
        id: Idx::<ExportItem>>(13),
        existing: Idx::<ExportItem>>(12),
    },
    DuplicateExport {
        id: Idx::<ExportItem>>(15),
        existing: Idx::<ExportItem>>(14),
    },
    DuplicateExport {
        id: Idx::<ExportItem>>(17),
        existing: Idx::<ExportItem>>(16),
    },
]
